### ISSUE TITLE
Generic constraints as return type in class methods

### ISSUE THREAD

**Issue by lirbank** [microsoft/TypeScript#35077]
**TypeScript Version:**  3.6.4 & 3.7.2 & 3.8.0-dev.20191112

**Search Terms:** generic constraints, TS2322, 2322

**Code**

```ts
export class Model<GenericSchema extends { id: string }> {
  public fails(): GenericSchema {
    return { id: '' };  // <-- This line produce the error, see below
  }

  public works(): GenericSchema {
    return { id: '' } as GenericSchema;
  }
}

```

**Expected behavior:**
Should compile without errors. To me this should work since the the `fails()` method returns an object which satisfies the constraint `{ id: string }`.

**Actual behavior:**
```
3:5 - error TS2322: Type '{ id: string; }' is not assignable to type 'GenericSchema'.
  '{ id: string; }' is assignable to the constraint of type 'GenericSchema', but 'GenericSchema' could be instantiated with a different subtype of constraint '{ id: string; }'.
```

**Context:**
For additional context, I originally filed an issue with `@types/mongodb` which gives the original use case
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/39358

Last, here is a discussion of the exact same matter on Stack Overflow
https://stackoverflow.com/questions/58663733/typescript-class-generic-constraint

**Playground Link:**
https://codesandbox.io/s/mapped-type-with-generics-l4b0b

**Related Issues:** 
I believe this is related to https://github.com/microsoft/TypeScript/issues/34567 and as such probably also https://github.com/microsoft/TypeScript/issues/33014.

**Comment by fatcerberus** [comment]
The error here is correct.  `Model` could be instantiated, for example, with `GenericConstraint = { id: string, foo: string }`, which is assignable to `{ id: string }`.  The inverse, however, is not true.

**Comment by AnyhowStep** [comment]
Here's a concrete example of the kind of error TS is trying to protect you from,

```ts
export class Model<GenericSchema extends { id: string }> {
  public fails(): GenericSchema {
    return { id: '' };  // <-- This line produce the error, see below
  }

  public works(): GenericSchema {
    return { id: '' } as GenericSchema;
  }
}
const model = new Model<{id:string, x:string}>();
const obj = model.works();
console.log(obj.x); //undefined
```

[Playground](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=12&pc=32#code/MYGwhgzhAECyD2ATApiAPAcWQO2QJwEtgBlYAC2QFsxpkAPAFx0RgG9oDEAuaCBw7AHNoAXwB80VgFgAUNGgAHAK4AjEEWgAzMARAQAFAEoeWXIRLkqNaXPnQ8yBkrzZJHbtADkn0QG55APQB0GgAtKHQACpkBDDquIp4SErAyNAMFLR4SXgANLzIaSqo8ADusvIishWKqurA0KXweADWBsbQpvhEpBTUkjXyDk4ubpw83qLQkJ043RZ9YL41VTKrwPDYfNCUSKjQALzQuKVwe+is43wCgvl0XNcEQuJGyzIbWwzQ8CoAVoc7c4AOiarXabw+EHgIGQQJA8EE+h+vyBdEM-iCSmwKE0T2QiCAA)

**Comment by lirbank** [comment]
Thanks a lot for the example, that makes a lot of sense! 

In my real use case I'm accessing a database so it would be OK for `obj.x ` to be out of sync, since this would be a developer error (providing the wrong type for the data model in the DB). That said, to me it now (with your help) seems the TS compile is handling the above example correctly.

Now, if I change the example to be a bit closer to a real use case with an external db driver module, it may look something like this. And now the compiler accepts the generic as a return type, without the type assertion - which also makes sense.

```ts
// TS don't know what's in the DB (obviously)
const dataInDb: any = {id: '', stuff: ''}

// DB Driver takes a generic which defines what is 
// in the DB (up to the developer to make sure the 
// generic match the data model) 
function readDb<Schema>():Schema {
  return dataInDb
}

class Model<Schema extends { id: string }> {
  // This now works without the type assertion
  public read(): Schema {
    return readDb<Schema>();
  }
}

const model = new Model<{id:string, x:string}>();
const obj = model.read();
console.log(obj.x); //undefined
```
[Playground](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=21&pc=1#code/PTAEBUGVQEwewHYHIAuoDWC4HdTYBYCGKSAzqAJYKgr4CmoAIgEKgAUcARgG4VwCupADYBPAJQBYAFABjRKTQxihAJIJGnAFyhCCEaAC8oAN4UY2pEgA0oBfwBm9i0gC+06SCatGAJwrc6HxpCdDpyQlAAczoEQIoZPHx4-Fg6eyowxOJKcg8wKhp6L3Z+AAcaOEKGGDoAoThSwIrQAFsQhlJ+HwZaBjyomLiEtpQZFN7YZVa4GqExUGl7fgQZFD5qbsIYDQAeSDG6NoA+NjFNffo2k2lQUG6ULuolFFV1Tmk3KWkZIUJScgAsjM6EI9gcrnQAB4oGIwcjGSjmWwoPwISKgFxHa5SW6ecBJchYXDYOA+dDkbAUWgCNATFAiRo6f6BNaIG6gUr8ThCeJ3OhbU7aC6HCLGdm3e6PPlbXbC46nADc7M+n2+8jQLWBQkMoFiuCBsx2pnMClRkRskM0pqokUxirVCAUoC4ACsdZrZgA6TYwe2yeRwIR0T31SIcTguz2QsQK0AgZY1dKxGDSIA)

A more contrived example:
```ts
class Model<GenericSchema extends { id: string }> {
  public fails(): GenericSchema {
    return { id: '' } as any;  // <-- This line does not produce any error anymore
  }

  public works(): GenericSchema {
    return { id: '' } as GenericSchema;
  }
}
const model = new Model<{id:string, x:string}>();
const obj = model.works();
console.log(obj.x); //undefined
```
[Playground](http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=12&pc=32#code/MYGwhgzhAECyD2ATApiAPAcWQO2QJwEtgBlYAC2QFsxpkAPAFx0RgG9oDEAuaCBw7AHNoAXwB80VgFgAUNGgAHAK4AjEEWgAzMARAQAFAEoeWXIRLkqNaXPnQ8yBkrzZJHbtADkn0dEh-sAE8AbnkAejDoNABaaOgAFTICGHVcaER4ZBhseAZFPCQlYGQAwNo8ArxSyngHWXkRWXrFVXVgaAB3WoBrA2NoU3wiUgpqSWb5BycXN04eb19-QfMRq2DmxplN4HhsPmgalBBoAF5oXA64JFQ0Vjm+AUEAGmg6LgeCIXEjdZkdvby8BUACtTgdriAAHRdPC9H6yf4QeAgZCQkDwQT6IHAyF0QyhCJKbAoTSfZCIIA)

In summary, my initial example was simply not correct and both examples works as expected. I'll close this issue and bring the matter back over to the `@types/mongodb` https://github.com/DefinitelyTyped/DefinitelyTyped/issues/39358 package.

Big thanks for your help @fatcerberus and @AnyhowStep!

**Comment by RyanCavanaugh** [comment]
https://twitter.com/SeaRyanC/status/1121995986862084096
